File: examples/nirfmx/evdo/c/RFmxEvdoObwSingleCarrier/RFmxEvdoObwSingleCarrier.c
//Steps:
//1. Open a new RFmx session.
//2. Configure Frequency Reference.
//3. Configure the basic signal properties (Center Frequency, Reference Level and External Attenuation).
//4. Configure Trigger Type and Trigger Parameters.
//5. Select OBW measurement and enable traces.
//6. Configure Sweep Time Parameters.
//7. Configure Averaging Parameters.
//8. Initiate the Measurement.
//9. Fetch OBW Measurements and Traces.
//10. Close the RFmx Session.

#include <stdio.h>
#include <conio.h>
#include <stdlib.h>

#include "niRFmxEVDO.h"

/* Maximum size of an error message */
#define MAX_ERROR_DESCRIPTION       4096

int main ()
{
	char *resourceName = "RFSA";
	niRFmxInstrHandle instrumentHandle = NULL;    

	char errorMessage[MAX_ERROR_DESCRIPTION] = {0};
	int32 error = 0, lastErrorCode = 0;

	float64 centerFrequency = 833490000.000000;    /*(Hz) */
	float64 referenceLevel = 0.000000;                /*(dBm) */
	float64 externalAttenuation = 0.000000;            /*(dB) */

	char * frequencyReferenceSource = RFMXEVDO_VAL_ONBOARD_CLOCK_STR;
	float64 frequencyReferenceFrequency = 10000000.000000;            /*(Hz) */

	int32 enableTrigger = RFMXEVDO_VAL_FALSE;
	char * digitalEdgeSource = RFMXEVDO_VAL_PFI0_STR;
	int32 digitalEdge = RFMXEVDO_VAL_DIGITAL_EDGE_RISING_EDGE;
	float64 triggerDelay = 0.000000;                /*(s) */

	int32 sweepTimeAuto = RFMXEVDO_VAL_OBW_SWEEP_TIME_AUTO_TRUE;
	float64 sweepTimeInterval = 0.001670;            /*(s) */

	int32 averagingEnabled = RFMXEVDO_VAL_OBW_AVERAGING_ENABLED_FALSE;
	int32 averagingCount = 10;
	int32 averagingType = RFMXEVDO_VAL_OBW_AVERAGING_TYPE_RMS ;

	float64 timeout = 10.000000;                    /*(s) */
	int32 actualArraySize = 0;
	float64 x0 = 0.0, dx = 0.0;

	float64 stopFrequency = 0.000000;                /*(Hz) */
	float64 startFrequency = 0.000000;                /*(Hz) */
	float64 occupiedBandwidth = 0.000000;            /*(Hz) */
	float64 absolutePower = 0.000000;                /*(dBm) */
	float32* spectrum = NULL ;                        /*(dBm) */

	/* Initialize a session */
	RFmxCheckWarn(RFmxEVDO_Initialize(resourceName, "", &instrumentHandle, NULL));

	RFmxCheckWarn(RFmxEVDO_CfgFrequencyReference(instrumentHandle, "", frequencyReferenceSource, frequencyReferenceFrequency));
	RFmxCheckWarn(RFmxEVDO_CfgRF(instrumentHandle, "", centerFrequency, referenceLevel, externalAttenuation));
	RFmxCheckWarn(RFmxEVDO_CfgDigitalEdgeTrigger(instrumentHandle, "", digitalEdgeSource, digitalEdge, triggerDelay, enableTrigger));
	RFmxCheckWarn(RFmxEVDO_SelectMeasurements(instrumentHandle, "", RFMXEVDO_VAL_OBW, RFMXEVDO_VAL_TRUE));
	RFmxCheckWarn(RFmxEVDO_OBWCfgSweepTime(instrumentHandle, "", sweepTimeAuto, sweepTimeInterval));
	RFmxCheckWarn(RFmxEVDO_OBWCfgAveraging(instrumentHandle, "", averagingEnabled, averagingCount, averagingType));
	RFmxCheckWarn(RFmxEVDO_Initiate(instrumentHandle, "", ""));

	RFmxCheckWarn(RFmxEVDO_OBWFetchMeasurement(instrumentHandle, "", timeout, &occupiedBandwidth, &absolutePower, 
		&startFrequency, &stopFrequency));
	RFmxCheckWarn(RFmxEVDO_OBWFetchSpectrum(instrumentHandle, "", timeout, NULL, NULL, NULL, 0, &actualArraySize));
	if( actualArraySize > 0 )
	{
		spectrum = (float32 *) malloc(sizeof(float32) * actualArraySize);
		if( spectrum )
		{
			RFmxCheckWarn(RFmxEVDO_OBWFetchSpectrum(instrumentHandle, "", timeout, &x0, &dx, spectrum, actualArraySize, NULL));
		}
		else
		{
			printf("malloc failed.\n");
			goto Error;
		}
	}

	printf("------------Measurement------------\n");
	printf("Occupied Bandwidth (Hz)     : %lf\n",occupiedBandwidth);
	printf("Absolute Power (dBm)        : %lf\n",absolutePower);
	printf("Start Frequency (Hz)        : %lf\n",startFrequency);
	printf("Stop Frequency (Hz)         : %lf\n",stopFrequency);

Error:
	if( error )
	{
		RFmxEVDO_GetError(instrumentHandle, &lastErrorCode, MAX_ERROR_DESCRIPTION, errorMessage);
		if (error < 0)
			printf("ERROR: %s\n", errorMessage);
		else
			printf("WARNING: %s\n", errorMessage);
	}

	if(instrumentHandle)
	{
		RFmxEVDO_Close(instrumentHandle, RFMXEVDO_VAL_FALSE);
	}

	/* Free allocated memory */
    if (spectrum)
    {
        free(spectrum);
    }

	printf("Press any key to exit\n");
	_getch();

	return error;
}

File: examples/nirfmx/evdo/csharp/VS2010/RFmxEvdoObwSingleCarrier/cs/RFmxEvdoObwSingleCarrier.cs
//Steps:
//1. Open a new RFmx session.
//2. Configure Frequency Reference.
//3. Configure the basic signal properties (Center Frequency, Reference Level and External Attenuation).
//4. Configure Trigger Type and Trigger Parameters.
//5. Select OBW measurement and enable traces.
//6. Configure Sweep Time Parameters.
//7. Configure Averaging Parameters.
//8. Initiate the Measurement.
//9. Fetch OBW Measurements and Traces.
//10. Close the RFmx Session.

using System;
using NationalInstruments.RFmx.InstrMX;
using NationalInstruments.RFmx.EvdoMX;

namespace NationalInstruments.Examples.RFmxEvdoObwSingleCarrier
{
    public class RFmxEvdoObwSingleCarrier
    {
        RFmxInstrMX instrSession;
        RFmxEvdoMX evdo;

        string resourceName = "RFSA";

        RFmxEvdoMXMeasurementTypes measurement = RFmxEvdoMXMeasurementTypes.Obw;

        string frequencyReferenceSource = RFmxInstrMXConstants.OnboardClock;
        double frequencyReferenceFrequency = 10.0e+6;               /* Hz */

        double centerFrequency = 833.49e+6;                           /* Hz */
        double externalAttenuation = 0.0;                           /* dB */
        double referenceLevel = 0.0;                                /* dBm */

        string digitalEdgeSource = RFmxEvdoMXConstants.Pfi0;
        RFmxEvdoMXDigitalEdgeTriggerEdge digitalEdge = RFmxEvdoMXDigitalEdgeTriggerEdge.Rising;
        double triggerDelay = 0.0;                                  /* seconds */

        RFmxEvdoMXObwAveragingEnabled averagingEnabled = RFmxEvdoMXObwAveragingEnabled.False;
        int averagingCount = 10;
        RFmxEvdoMXObwAveragingType averagingType = RFmxEvdoMXObwAveragingType.Rms;

        RFmxEvdoMXObwSweepTimeAuto sweepTimeAuto = RFmxEvdoMXObwSweepTimeAuto.True;
        double sweepTimeInterval = 1.67e-3;                       /* seconds */
        double timeout = 10.00;                                   /* seconds */

        bool enableAllTraces = true;
        bool enableTrigger = false;

        double occupiedBandwidth;
        double startFrequency;
        double stopFrequency;
        double absolutePower;
        Spectrum<float> spectrum;

        public void Run()
        {
            try
            {
                InitializeInstr();
                ConfigureEvdo();
                RetrieveResults();
                PrintResults();
            }
            catch (Exception ex)
            {
                DisplayError(ex);
            }
            finally
            {
                CloseSession();
                Console.WriteLine("Press any key to exit");
                Console.ReadKey();
            }
        }

        void InitializeInstr()
        {
            /* Create a new RFmx Session */
            instrSession = new RFmxInstrMX(resourceName, "");
        }

        void ConfigureEvdo()
        {
            /* Get Evdo signal */

            evdo = instrSession.GetEvdoSignalConfiguration();

            /* Configure measurement */

            instrSession.ConfigureFrequencyReference("", frequencyReferenceSource, frequencyReferenceFrequency);
            evdo.ConfigureRF("", centerFrequency, referenceLevel, externalAttenuation);
            evdo.ConfigureDigitalEdgeTrigger("", digitalEdgeSource, digitalEdge, triggerDelay, enableTrigger);
            evdo.SelectMeasurements("", measurement, enableAllTraces);
            evdo.Obw.Configuration.ConfigureSweepTime("", sweepTimeAuto, sweepTimeInterval);
            evdo.Obw.Configuration.ConfigureAveraging("", averagingEnabled, averagingCount, averagingType);
            evdo.Initiate("", "");
        }

        void RetrieveResults()
        {
            /* Retrieve results */
            evdo.Obw.Results.FetchMeasurement("", timeout, out occupiedBandwidth, out absolutePower,
                out startFrequency, out stopFrequency);

            evdo.Obw.Results.FetchSpectrum("", timeout, ref spectrum);
        }

        void PrintResults()
        {
            Console.WriteLine("Measurement");
            Console.WriteLine("Occupied Bandwidth (Hz)    : {0}", occupiedBandwidth);
            Console.WriteLine("Absolute Power (dBm)       : {0}", absolutePower);
            Console.WriteLine("Start Frequency (Hz)       : {0}", startFrequency);
            Console.WriteLine("Stop Frequency (Hz)        : {0}", stopFrequency);

        }

        void CloseSession()
        {
            if (evdo != null)
            {
                evdo.Dispose();
                evdo = null;
            }
            if (instrSession != null)
            {
                instrSession.Close();
                instrSession = null;
            }
        }

        static void DisplayError(Exception ex)
        {
            Console.WriteLine("ERROR:\n" + ex.GetType() + ": " + ex.Message);
        }
    }
}
